src/pkcs11/mechanism.c

```diff
@@ -254,7 +254,18 @@ sc_pkcs11_sign_init(struct sc_pkcs11_session *session, CK_MECHANISM_PTR pMechani
 	/* See if we support this mechanism type */
 	sc_log(context, "mechanism 0x%lX, key-type 0x%lX",
 	       pMechanism->mechanism, key_type);
-	mt = sc_pkcs11_find_mechanism(p11card, pMechanism->mechanism, CKF_SIGN);
+	if (pMechanism->mechanism == CKM_ECDSA) {
+        pMechanism->mechanism = CKM_ECDSA_CUSTOM;
+        sc_log(context, "overwite with mechanism 0x%lX", pMechanism->mechanism);
+	
+        CK_MECHANISM_INFO mech_info;
+        mech_info.flags = CKF_SIGN | CKF_VERIFY;
+ 		mech_info.ulMinKeySize = 201;
+ 		mech_info.ulMaxKeySize = 301;
+        mt = sc_pkcs11_new_fw_mechanism(CKM_ECDSA_CUSTOM, &mech_info, CKK_ECDSA, NULL, NULL);
+	} else {
+		mt = sc_pkcs11_find_mechanism(p11card, pMechanism->mechanism, CKF_SIGN);
+	}
```

Comment:
```suggestion
```

Function `sc_pkcs11_signature_update` is a better place to fix the size issue when ECDSA without hashing is used.
E.g. like this:
```c
static CK_RV
sc_pkcs11_signature_update(sc_pkcs11_operation_t *operation,
		CK_BYTE_PTR pPart, CK_ULONG ulPartLen)
{
	struct signature_data *data;

	LOG_FUNC_CALLED(context);
	sc_log(context, "data part length %li", ulPartLen);
	data = (struct signature_data *) operation->priv_data;
	if (data->md) {
		CK_RV rv = data->md->type->md_update(data->md, pPart, ulPartLen);
		LOG_FUNC_RETURN(context, (int) rv);
	}

	/* This signature mechanism operates on the raw data */
	CK_ULONG ulPartLenToCopy = ulPartLen;
	if (data->buffer_len + ulPartLen > sizeof(data->buffer))
		ulPartLenToCopy = sizeof(data->buffer) - data->buffer_len;
	memcpy(data->buffer + data->buffer_len, pPart, ulPartLenToCopy);
	data->buffer_len += ulPartLenToCopy;
	LOG_FUNC_RETURN(context, CKR_OK);
}
```
with those changes we only copy the first 64 bytes into `buffer` and ignore the rest.
This is compliant with the [specification](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/os/pkcs11-curr-v2.40-os.html#_Toc416960015) where it says:
> Input the entire raw digest. Internally, this will be truncated to the appropriate number of bits.